examples/validate-fee/modules/shared_data/src/main/scala/com/my/validate_fee/shared_data/calculated_state/DataCalculatedStateService.scala

```diff
@@ -0,0 +1,42 @@
+package com.my.validate_fee.shared_data.calculated_state
+
+import cats.Monoid
+import cats.effect.Ref
+import cats.effect.Async
+import cats.syntax.all._
+import io.circe.Encoder
+import io.circe.syntax.EncoderOps
+import org.tessellation.currency.dataApplication.DataCalculatedState
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.security.hash.Hash
+
+import java.nio.charset.StandardCharsets
+
+trait DataCalculatedStateService[F[_], A <: DataCalculatedState] {
+  def getCalculatedState: F[(SnapshotOrdinal, A)]
+
+  def setCalculatedState(snapshotOrdinal: SnapshotOrdinal, state: A): F[Boolean]
+
+  def hashCalculatedState(state: A): F[Hash]
+}
+
+object DataCalculatedStateService {
+
+  def make[F[_] : Async, A <: DataCalculatedState : Monoid: Encoder]: F[DataCalculatedStateService[F, A]] = {
+    Ref.of[F, (SnapshotOrdinal, A)](SnapshotOrdinal.MinValue -> Monoid[A].empty).map { stateRef =>
+      new DataCalculatedStateService[F, A] {
+        override def getCalculatedState: F[(SnapshotOrdinal, A)] = stateRef.get
+
+        override def setCalculatedState(snapshotOrdinal: SnapshotOrdinal, state: A): F[Boolean] =
+          stateRef.modify { case (_, currentState) =>
+            ((snapshotOrdinal, currentState |+| state), true)
```

Comment:
I'm not sure if combining the current state and previous state here is advisable. It seems like `DataApplicationSnapshotAcceptanceManager.consumeSignedMajorityArtifact` is the location where this set function is modified (other than download operations where this is just set). Based on this portion of the function, then it seems like if the `CalculatedState` is modified correctly in the Combiner functions then shouldn't state already be a super set (or contain the data) that is within current state?

```scala
da =>
          OptionT
            .liftF(da.blocks.traverse(service.deserializeBlock).map(_.flatMap(_.toOption)))
            .flatMapF { dataBlocks =>
              artifact.ordinal.partialPrevious.flatTraverse(lastOrdinal =>
                accept(maybeLastDataApplication, dataBlocks, lastOrdinal, artifact.ordinal)
              )
            }
            .map(_.calculatedState)
            .semiflatMap(expectCalculatedStateHash(da.calculatedStateProof))
            .semiflatTap(service.setCalculatedState(artifact.ordinal, _))
            .semiflatTap(calculatedStateStorage.write(artifact.ordinal, _)(service.serializeCalculatedState))
```

I believe this function can be simplified to 
```scala
          override def set(checkpoint: Checkpoint[S]): F[Boolean] =
            state
              .set(checkpoint)
              .as(true)
              .handleErrorWith { err =>
                Logger[F].warn(s"Checkpoint set failed with error: $err") *> false.pure[F]
              }
```

Reply:
@scasplte2 If understand correctly, are you suggesting `setCalculatedState` simply overwrite the existing state with the one given?

Reply:
yes, the combiner takes care of generating the new state that is associated to the ordinal so I believe it is redundant to do it again here.